core/src/main/java/org/prevayler/TransactionBase.java

```diff
@@ -0,0 +1,19 @@
+//Prevayler(TM) - The Free-Software Prevalence Layer.
+//Copyright (C) 2001-2003 Klaus Wuestefeld
+//This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+package org.prevayler;
+
+import java.io.Serializable;
+
+/**
+ * Formal common base type for {@link Transaction} and {@link TransactionWithQuery}.
+ * Used to limit the valid types (via ...<code>? extends TransactionBase</code>...)
+ * in classes/methods/parameters referring to either of the two types using generics
+ *      
+ * @author Christian Klemke
+ * @see Transaction
+ * @see TransactionWithQuery
+ */
+public interface TransactionBase extends Serializable {
```

Comment:
This is nice. It definitely improves some of the `Object` usages in the internals. I'm a tiny bit concerned that it's visible in the public API even though it's purely for internal code tidiness, but after giving it some thought I think that I'm okay with it.

Reply:
I understand your concern, but I feel it is still "sufficiently" elegant like this compared to previous implementation. Besides, it more type-safe. Actually, I cannot think of any other way to make Transaction and TransactionWithQuery be usable in the corresponding places in an exchangeable way without casts and without losing type safety.